SparseGrids/TasmanianSparseGrid.cpp

```diff
@@ -391,16 +391,30 @@ void TasmanianSparseGrid::getDifferentiationWeights(const std::vector<double> &x
     getDifferentiationWeights(x.data(), weights.data());
 }
 void TasmanianSparseGrid::getDifferentiationWeights(const double x[], double weights[]) const{
+    // See differentiate() for how the domain transforms are taken into consideration.
     Data2D<double> x_tmp;
+    // Jacobian of f(.) at g(x).
     if (isGlobal()) {
         get<GridGlobal>()->getDifferentiationWeights(formCanonicalPoints(x, x_tmp, 1), weights);
     } else if (isSequence()) {
         get<GridSequence>()->getDifferentiationWeights(formCanonicalPoints(x, x_tmp, 1), weights);
     } else if (isLocalPolynomial()) {
         get<GridLocalPolynomial>()->getDifferentiationWeights(formCanonicalPoints(x, x_tmp, 1), weights);
+    } else if (isFourier()) {
+        get<GridFourier>()->getDifferentiationWeights(formCanonicalPoints(x, x_tmp, 1), weights);
     } else {
         throw std::runtime_error("ERROR: getDifferentiationWeights() cannot be called for grids of this type");
     }
+    // Jacobian of f(g(.)) at x.
+    if (domain_transform_a.size() != 0 or conformal_asin_power.size() != 0) {
```

Comment:
If you have unfinished features like the domain transforms, throw an exception and remove it late. Right now, someone may use this an get garbage results not realizing what's happening.

Reply:
This is handled by the new `TasmanianSparseGrid::diffCanonicalTransform()` function in this PR. Specifically, that function throws an exception if conformal transforms are in place, but not affine ones.

Reply:
Why is the if-statement needed here?

Reply:
The if-statement is needed to avoid calling `TasmanianSparseGrid::diffCanonicalTransform()` when the available transformation data, e.g., `domain_transform_a`, is empty, since that function requires it to be non-empty. We also save on some allocations and excessive operations.

Reply:
Read the code inside `diffCanonicalTransform()`, you are first checking if `conformal_asin_power` is not empty (and raising an exception), then checking if `domain_transform_a` is not empty and doing things. This if-statement looks totally redundant.

Reply:
On a related note, `conformal_asin_power.size() != 0` works, but sometimes it is more expressive to use
```
if (not domain_transform_a.empty() or not conformal_asin_power.empty())
```

Reply:
Ah, I missed that. In that case, I have changed the if-statement to 
```
if (not domain_transform_a.empty())
```
since we want to avoid the allocation in `TasmanianSparseGrid::diffCanonicalTransform()` and the expensive double for-loop in the if-block when `domain_transform_a` is empty, i.e., when we don't need to scale the canonical Jacobian.